web/api/v1/api.go

```diff
@@ -521,28 +520,13 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
 		return apiFuncResult{nil, &apiError{errorBadData, errors.Wrap(err, "invalid parameter 'end'")}, nil, nil}
 	}
 
-	q, err := api.Queryable.Querier(r.Context(), timestamp.FromTime(start), timestamp.FromTime(end))
-	if err != nil {
-		return apiFuncResult{nil, &apiError{errorExec, err}, nil, nil}
-	}
-	// From now on, we must only return with a finalizer in the result (to
-	// be called by the caller) or call q.Close ourselves (which is required
-	// in the case of a panic).
-	defer func() {
-		if result.finalizer == nil {
-			q.Close()
-		}
-	}()
-	closer := func() {
-		q.Close()
-	}
+	vals, warnings, err := api.webkit.labelValues(r.Context(), name, start, end)
```

Comment:
In all these cases, I haven't double-checked all the logic around this finalizer stuff yet, especially around where a panic can happen (what is outlined in the removed comments). There seems to be a reason why we don't usually close the querier directly in the handler functions, but pass it up to the caller and let it handle it at the end? Something to double-check at the end...

Reply:
For this part, to be honest I didn't understand why it was like that. Because in all cases the finalizer is the same.
And the `defer` function is called even during a panic. Or to do so, it should be wrapped in a anonymous function like it was the case before. Maybe that's the only thing that should be modified here.

But maybe it required some golang knowledge I don't have.

Reply:
This isn't a Go thing, it's a Prometheus/TSDB thing. Any strings returned are only valid until the finaliser is called, as they could be yoloStrings backed by mmap.

Reply:
So if I understood correctly, if it closed too earlier, the string is going to be deleted ?

Sorry in advance for the dummy question, but I thought the string type cannot be changed like that unless you are using a pointer. And since it doesn't return a pointer, then how the result can be changed once the finaliser is called ?

Reply:
https://github.com/prometheus/prometheus/blob/0cff72d756d3b0eb1690dce66b9ca62b32f1deb8/tsdb/index/index.go#L1721-L1723

Reply:
ok thanks for the answer. I think I understand better why you would like to avoid the introduction of the `webkit` struct @brian-brazil.

I will try to propose something easy to handle this closure issue.

Reply:
but then if there is a problem here, then we will have the same problem for the langServer endpoints. So the endpoints won't return anything.

Reply:
or to avoid this issue, I could copy the result of the query, so it will be immutable and then the closure can stay in the `defer` statement and in the method associated to the struct `webkit`

Reply:
That's one approach, though it'll create more garbage.

Reply:
yeah you are right, that's not a good approach. I didn't have the mind clear lol.
I need to think more about it.

Reply:
So here I choose to return the query created. Like that the API can close it when needed.

On the langserver side, I had to copy the result before closing the query

I hope you are ok with this approach